scripts/data/client.py

```diff
@@ -152,52 +155,49 @@ def job_producer(job_gen):
                     and current_weight != 0
                     or job_queue.full()
                 ):
+                    if shutdown_event.is_set():
+                        logger.info("Shutdown event detected while waiting. Exiting...")
+                        return
+
                     logger.debug("Producer is waiting for weight to be released.")
-                    weight_lock.wait()  # Wait for the condition to be met
+                    weight_lock.wait()
```

Comment:
What will happen when SIGTERM is sent when thread is waiting for a lock?

Reply:
Threads waiting on the lock will wait until they are notified (either by a timeout or by other threads) before they can check the `shutdown_event` and exit.

Reply:
This means SIGTERM will not terminate the program immediately?

Reply:
Yes, it will not terminate the program immediately rather it initiates a graceful shutdown process which will allow active threads to finish their current tasks and lets waiting threads proceed once they acquire the lock.

Reply:
The point of sending SIGTERM is to stop immediately.